src/channel_internal.c

```diff
@@ -84,20 +84,20 @@ static void abandon_pending_operations(void* context)
         /*Codes_SRS_CHANNEL_INTERNAL_43_168: [abandon_pending_operations shall set is_open to false.]*/
         channel_internal_ptr->is_open = false;
 
+        /*Codes_SRS_CHANNEL_INTERNAL_43_169: [ abandon_pending_operations shall call srw_lock_release_exclusive. ]*/
+        srw_lock_release_exclusive(channel_internal_ptr->lock);
+    }
 
-        /*Codes_SRS_CHANNEL_INTERNAL_43_095: [ abandon_pending_operations shall iterate over the list of pending operations and do the following: ]*/
-        for (DLIST_ENTRY* entry = DList_RemoveHeadList(&channel_internal_ptr->op_list); entry != &channel_internal_ptr->op_list; entry = DList_RemoveHeadList(&channel_internal_ptr->op_list))
-        {
-            CHANNEL_OP* channel_op = CONTAINING_RECORD(entry, CHANNEL_OP, anchor);
+    /*Codes_SRS_CHANNEL_INTERNAL_43_095: [ abandon_pending_operations shall iterate over the list of pending operations and do the following: ]*/
+    for (DLIST_ENTRY* entry = DList_RemoveHeadList(&channel_internal_ptr->op_list); entry != &channel_internal_ptr->op_list; entry = DList_RemoveHeadList(&channel_internal_ptr->op_list))
```

Comment:
>for (DLIST_ENTRY* entry = DList_RemoveHeadList(&channel_internal_ptr->op_list); entry != &channel_internal_ptr->op_list; entry = DList_RemoveHeadList(&channel_internal_ptr->op_list)) [](http://example.com/codeflow?start=4&length=182)

Now you're using the DLIST outside of the lock.  This _might_ be safe, but it is very fragile. :confused:

Reply:
I say very fragile because using a lock to protect op_list in all cases except one is the same as not using a lock.  There is now a possible code path where op_list gets corrupted by multithreaded access.  You can try to imagine every possible interleaving of code to prove that using `is_open` is good enough in this case, but:
1. Relying on is_open is never going to be as good as using a lock.
2. The results of that exercise are invalidated as soon as the code changes.

Reply:
In the next iteration, I made the change to make a local copy under lock and nulling out the original list.